### PR TITLE
feat(watcher): symlink/containment hardening for file events (WP-2.3, #1228)

### PR DESCRIPTION
**Pull request by curdriceaurora**
## WP-2.3 — Watcher event symlink/containment hardening

Part of epic #1221 (fo-core pull-back). Closes #1228 (fo-core#322).

### Problem

`watcher.handler.FileEventHandler` enqueued **every** event that passed the
exclude/whitelist filter, with no symlink or containment check. A symlink
planted in the watched tree — or a `MOVED` event whose destination escapes the
watched root — would drive a downstream read/organize of out-of-root content
(the same symlink-exfiltration surface WP-2.1/WP-2.2 closed at the read and
write boundaries, now closed at the event-ingest boundary).

### Change

- **`handler.py`** — new fail-closed guard `_is_event_path_allowed` (+
  `_resolved_watch_roots`), invoked in `_handle_event` before debounce/enqueue:
  - final-component symlink refused via `os.lstat` → `stat.S_ISLNK`
  - path canonicalized with `Path.resolve()`; must be inside a resolved watch root
  - any resolution error (symlink loop, OS error) → treated as unsafe, skipped
  - `MOVED` destination is checked too (it's the path that gets organized)
  - inert when no watch roots are configured (standalone handler keeps working)
- **`monitor.py`** — `add_directory` now keeps `config.watch_directories` in sync
  when a directory is added **mid-run**. Previously the running branch scheduled
  the observer watch without updating config, so the new guard would skip events
  for directories added after `start()`. This also means a later restart
  re-watches the dynamically-added directory.

### Tests

`tests/watcher/test_safedir_hardening_322.py` (9 cases, `unit`+`ci` marked):
in-root pass, no-roots-inert, out-of-root skip, symlinked leaf / symlinked
ancestor / symlink loop refusal, move-dest containment, move-within-root pass,
and a deterministic monkeypatched resolve-failure fail-closed case.

Full `tests/watcher/` suite: **146 passed**. ruff/format/mypy(changed)/diff-cover
80%/PyMarkdown all clean.

https://claude.ai/code/session_01L4urW7VLML2ReqxaHRc2Br

---
_Generated by [Claude Code](https://claude.ai/code/session_01L4urW7VLML2ReqxaHRc2Br)_

## Release Notes

* **Security**
  * Hardened file watcher event handling with fail-closed path containment checks and per-component symlink detection.
  * Event processing now rejects unsafe paths, including resolution/stat failures and symlink escape scenarios.
  * For move events, filtering and processing now prefer the destination path when available.

* **Improvements**
  * Dynamic directory additions during active monitoring now immediately synchronize the watched directory configuration (with rollback on scheduling failure).

* **Tests**
  * Expanded watcher hardening test coverage for edge cases, including canonicalization failures and move/destination boundary enforcement.

### CODE REVIEW

File: src/file_organizer/watcher/handler.py [hunk 1]
```diff
@@ -133,16 +134,33 @@ def _handle_event(
...
         if not is_directory and not self.config.should_include_file(path):
             logger.debug("Filtered out event for: %s", path)
             return
 
+        # Symlink / containment hardening (WP-2.3, #1228): refuse events whose
+        # path is a symlink (at any component under the watched root) or resolves
+        # outside the watched roots, failing closed on any resolution error
+        # (e.g. symlink loops).
+        if not self._is_event_path_allowed(
+            path, allow_missing_leaf=event_type is EventType.DELETED
```
[the remark this answers is not in the document]
Comment (curdriceaurora) [reply src/file_organizer/watcher/handler.py:159]: Valid — this is the irreducible check-at-enqueue vs open-at-use TOCTOU: the handler validates at enqueue time, but `process_file(event.path)` re-follows a symlink swapped in afterward. Closing it requires binding the read to the boundary at open time (anchored `SafeDir`/`O_NOFOLLOW`) in the orchestrator watch-mode read path, which is a separate surface from the WP-2.3 watcher event-ingest guard delivered here. Tracked as #1273 to keep this PR scoped. The handler guard remains valuable defense-in-depth (rejects symlink/out-of-root paths at ingest), and the authoritative no-follow open is the follow-up.

---
_Generated by [Claude Code](https://claude.ai/code)_
